crates/ursa-network/src/tests/service_tests.rs

```diff
@@ -191,42 +189,71 @@ mod tests {
-                    if let SwarmEvent::Behaviour(BehaviourEvent::Discovery(DiscoveryEvent::Connected(peer_id))) = event_2 {
+                event = node_1.swarm.select_next_some() => node_1.handle_swarm_event(event)?,
+                event = node_2.swarm.select_next_some() => match event {
+                    SwarmEvent::ConnectionEstablished { peer_id, ..} => {
+                        node_2.handle_swarm_event(event)?;
                         info!("[SwarmEvent::ConnectionEstablished]: {peer_id:?}, {peer_id_1:?}");
                         if peer_id == peer_id_1 {
                             break
                         }
-                    }
-                }
-                _ = node_1.swarm.select_next_some() => {}
+                    },
+                    event => node_2.handle_swarm_event(event)?,
+                },
             }
         }
         Ok(())
     }
 
     #[tokio::test]
-    async fn test_network_discovery() -> Result<()> {
+    async fn test_network_kad() -> Result<()> {
         setup_logger(LevelFilter::Info);
-        let mut config = NetworkConfig {
-            mdns: true,
-            ..Default::default()
-        };
 
-        let (mut node_1, _, peer_id_1, ..) = network_init(&mut config, None, None).await?;
-        let (mut node_2, ..) = network_init(&mut config, None, None).await?;
+        let (bootstrap, bootstrap_addr, bootstrap_id) =
+            run_bootstrap(&mut NetworkConfig::default()).await?;
 
+        // no need to loop over bootstraps events, we can just start it up
+        tokio::task::spawn(async move { bootstrap.start().await.unwrap() });
+
+        let (mut node_1, _, peer_id_1, ..) = network_init(
+            &mut NetworkConfig::default(),
+            Some(bootstrap_addr.clone()),
+            None,
+        )
+        .await?;
+
+        // wait for node 1 to identify with bootstrap
         loop {
             select! {
-                event_2 = node_2.swarm.select_next_some() => {
-                    if let SwarmEvent::Behaviour(BehaviourEvent::Discovery(DiscoveryEvent::Connected(peer_id))) = event_2 {
+                event = node_1.swarm.select_next_some() => match event {
+                    SwarmEvent::Behaviour(BehaviourEvent::Identify(libp2p::identify::Event::Received {peer_id, ..})) => {
+                        info!("Node 1 identified with bootstrap {bootstrap_id:?}");
+                        if peer_id == bootstrap_id {
+                            break
+                        }
+                    }
+                    event => node_1.handle_swarm_event(event)?,
+                },
+            }
+        }
+
+        let (mut node_2, ..) =
+            network_init(&mut NetworkConfig::default(), Some(bootstrap_addr), None).await?;
+
+        // wait for node 2 to connect with node 1 through kad peer discovery
+        loop {
+            select! {
+                event = node_1.swarm.select_next_some() => node_1.handle_swarm_event(event)?,
```

Comment:
Let’s in-line the event handler here for visibility

Reply:
Not sure what you mean?

Reply:
Like similar to the other tests where we are explicit about the event being handled

Reply:
We're not looking for events in node 1, only node 2, so we just pass to the service handler to let the node do it's thing in the background. Otherwise, the mdns events aren't handled and the nodes don't dial eachother

Reply:
Same comment still applies, instead of calling the handle swarm events, we should inline for visibility and consistency with other tests

Reply:
other tests just discard the event, and none of the service.rs logic is actually used. With mdns if you just disable bootstrap nodes and don't make any other modifications, it'd loop endlessly because the mdns handler never gets the event so the dial never happens etc. Since the refactor in #96 it's been doing this because the internal behavior poll() doesn't execute the code. Also some tests still work because start() is used, which loops and passes event to swarm_event_handler

Reply:
I am not sure I am following. When you call start the event loop gets started. Which is what should happen. My focus is on inclining the event handler and not using the handle_event function

Reply:
Yes, since we're selecting events before calling start, the actual event loop that calls handle_event is not present. And since kad and mdns logic is now included in the service.rs, it doesn't work unless you explicitly allow it to

Reply:
I am not sure you have a full picture of how the event loop works. I'll suggest a change once I am back on my laptop

Reply:
Sure thing, but just look at tests on main now, had to cancel because endless loop as events aren't being handled. I manually debugged and verified this earlier today

Reply:
For discovery?

Reply:
Event handling is best to be handled inlined for visibility

Reply:
yeah both kad and mdns are endless looping

Reply:
Mdns takes time to be handled, discovery can offer false positives.

Reply:
Maybe a util method that wraps select_next_some() and passes to handler, and then returns the event, instead of just returning next event would be a good improvement. Then start could use it, and we could use it in tests instead of select_next_some to keep the code less verbose

Reply:
We are not looking to rewrite tests atm, but could be a good future improvement

Reply:
Does this resolve what you wanted? https://github.com/fleek-network/ursa/pull/184/commits/77d4891d1759815804c13da8f0933d1979913018

Reply:
https://github.com/fleek-network/ursa/pull/184/commits/c84f5e1a9ec7ddd4ec00d98fd5571b2fbeb942f6